택배송장앱 [STEP 1] hjlee22

HJLEE-22 opened the pull request:
안녕하세요, 택배송장앱 STEP1 PR 드립니다.  @uuu1101

HJLEE-22 committed:
객체미용체조 7원칙 ‘2개 이상의 원시타입 프로퍼티를 갖는 타입 금지

- ParcelInformation의 address, name, mobile 세 가지 원시타입 프로퍼티를 사용하는 변수들을 ReceiverInformation 타입으로 치환해 원시타입 방지

HJLEE-22 committed:
SOLID SRP 적용 및 SOLID DIP 적용

- 단일책임원칙을 위해 DatabaseParcelInformationPersistence 타입 만들어 기능 분리
- 의존성 분리 원칙을 위해 ParcelInformationPersistence 프로토콜을 만들고, 해당 프로토콜을 소유하도록 변경

HJLEE-22 committed:
SOLID DIP 수정

- 의존성 주입 타입을 새로 생성한 ParcelInformationPersistence 프로토콜 타입으로 정의하도록 변경
- 따라서 DatabaseParcelInformationPersistence 외에도 ParcelInformationPersistence 프로토콜을 채택한 다른 타입들도 적용가능하도록 호완성 증대

HJLEE-22 commented:
두번째 커밋에서 SOLID SRP 적용 후 바로 DIP 적용 단계로 넘어가 함께 커밋한 부분 양해 부탁드립니다.

uuu1101 reviewed (approved):
안녕하세요 @HJLEE-22 🙇🏻
리팩토링 과정동안 잘 부탁드립니다.

첫 STEP을 잘 작성해주셨네요  !
코드량이 많지 않고 스텝에서 요구하는 부분들을 잘 작성해주셔서, 피드백을 드릴만한 내용이 크게 없을것 같습니다 ! 

진행하시면서 고민점이 있으시다면 PR에 같이 올려주시면 함께 고민해보아요 ! 
고생하셨습니다 👍🏻

uuu1101 commented:
다음 스텝을 진행하시기 전에 ParcelOrderViewController가 갖는 의존성 역전에 대한 부분도 리팩토링 해보시면 좋을것 같습니다.
다음 스텝 진행하시면서 하셔도 무방합니다. 어떻게 하실건지만 코멘트 남겨주세요 ! 😀

HJLEE-22 committed:
SOLID OCP 적용

- 개방폐쇠원칙 적용을 위해 계산 프로퍼티 내부의 switch 구현부분을 각각의 별도 타입으로 분리, 이후 Discount 타입에서 함께 관리
- ParcelInformation 클래스에서 해당 반환값을 불러오기 위해 지연 프로퍼티로 생성, 클래스 내부의 다른 값들을 사용

HJLEE-22 commented:
안녕하세요, 답변이 늦었습니다.
우선 택배송장앱 Step2 먼저 pr 드렸습니다. 
말씀해주신 ParcelOrderViewController가 갖는 의존성 역전에 대한 부분 리팩토링 관련도 오늘 안으로 업로드해 보겠습니다.
리뷰 감사드립니다.

HJLEE-22 committed:
ParcelOrderViewController 의존성 역전 원칙 적용

- ParcelOrderViewController에서 parcelProcessor 객체를 의존하지 않도록 별도 프로토콜 타입 생성
- 해당 프로토콜을 만족하는 객체를 생성자에서 주입받도록 정의

uuu1101 reviewed (approved):
@HJLEE-22 같은 PR에 올려주셔서 확인이 조금 늦었네요 ! 
STEP2도 잘 구현해주셨네요 ! 
> 그런데, 이렇게 계속 타입을 새로 만들고 의존성을 없에다보니 코드 가독성이 너무 떨어지는 것 아닌가 하는 의문이 듭니다.
- 맞습니다. 너무 추상화를 많이 하게 되면 가독성이 오히려 안좋아질 수 있습니다.
하지만 설계를 할때 계층을 잘 나누고 추상화를 하게 되면 개인적으로는 가독성이나 코드를 관리하는데 있어서 더 좋은 경험이 많았던것 같습니다 ! 

코드 내용이 많지 않아서 따로 드릴 만한 코드 관련 피드백은 크게 없고 진행하시면서 고민점이 있으시다면 언제든 말씀해주세요. 고생하셨습니다 👍🏻👍🏻
